Merge develop Branch into main

Atabek04 opened the pull request:
This pull request aims to merge the develop branch into the main branch. It includes the initial project setup and the implementation of core features for the Gym CRM application.

#### Summary of Changes:
-   Project Setup:
    - Initial Spring Boot project configuration.
    - Creation of base packages and directory structure.
    - Configuration files and dependencies setup.
- Core Features:
  - Trainee Management: CRUD operations for managing trainee profiles.
  - Trainer Management: CRUD operations for managing trainer profiles.
  - Training Management: CRUD operations for managing training sessions.
  - User Management: User authentication and management functionalities.
  - Command-Line Interface (CLI): Interactive CLI for managing the application.
- Additional Enhancements:
  - Implementation of mappers for converting DTOs and entities.
  - Integration of validation and error handling mechanisms.
-   Unit tests adhering to the F.I.R.S.T. principles and the AAA pattern for all services and DAOs.

#### Request for Review:
- Code Quality: Reviewers are requested to check for adherence to best practices, such as SOLID principles, clean code, and efficient design patterns.
- Functionality: Ensure all CRUD operations and CLI functionalities work as expected.
- Testing: Validate the adequacy and coverage of unit tests.

Please review and provide feedback or approval for merging into main.

MadinaBaizhanova commented on src/main/java/com/gymcrm/facade/GymFacade.java:
```diff
@@ -0,0 +1,146 @@
+        return userService.getAllUsers();
+    }
+
+    public User convertToUser(TraineeRequest request) {
```
Consider moving the call of the generateUsername() and generatePassword() methods to the Service layer. Per my understanding, this part can be considered as a business logic and should be in the Service layer, not in the Controller/Facade. That's just my subjective opinion.

MadinaBaizhanova commented on src/main/java/com/gymcrm/dto/TrainingResponse.java:
```diff
@@ -0,0 +1,40 @@
+@Value
+@Builder
+public class TrainingResponse {
+    int id;
```
in the future, consider replacing int by BigInteger for id's data type.

Atabek04 replied:
You're right! I was just certain that the number of entities will remain within the Integer range, which is of course less common in scalable, modern applications.

MadinaBaizhanova commented on src/main/java/com/gymcrm/dto/TraineeRequest.java:
```diff
@@ -0,0 +1,35 @@
+@Builder
+@AllArgsConstructor(access = AccessLevel.PUBLIC)
+public class TraineeRequest {
+    @NotNull @Positive int traineeId;
```
I am not sure that these annotations work. I have just managed to create a Trainee with id being -100 and -200. Also, I managed to create a Trainee with no/empty first and last names.

![image](https://github.com/user-attachments/assets/0cb991f0-e30c-4fe1-b223-c331be316f9e)

MadinaBaizhanova commented on src/main/java/com/gymcrm/dto/TrainerRequest.java:
```diff
@@ -0,0 +1,33 @@
+@Builder
+@AllArgsConstructor(access = AccessLevel.PUBLIC)
+public class TrainerRequest {
+    @Positive
```
Please check that the annotations do work so that when no first name or last name are provided, the corresponding messages from the validation annotations are shown

MadinaBaizhanova replied:
@Atabek04 , as I understand these annotations do not actually work. For example, I just managed to create a Trainee and with first and last names as 'f' and 'f' resulting in a username of 'f.f'.

So, I think that these annotation should either be removed at all or should be forced to actually work.

MadinaBaizhanova commented on src/main/java/com/gymcrm/mapper/TraineeMapper.java:
```diff
@@ -0,0 +1,31 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class TraineeMapper {
```
Since this is a utility class, consider making sure that no other classes can extend or instantiate this TraineeMapper class.

Atabek04 replied:
You're right! But in order to make code more cleaner, I'm using import static, and for that I need to class to be not final.
```java
import static com.gymcrm.mapper.TrainingMapper.toTraining;
```
```java
public void saveTraining(TrainingRequest trainingRequest) {
        var training = toTraining(trainingRequest);
        trainingService.createTraining(training);
    }
```

MadinaBaizhanova replied:
As I understand 'final' only prevents inheritance, it does allow other classes to use its static methods

MadinaBaizhanova commented on src/main/java/com/gymcrm/mapper/TrainerMapper.java:
```diff
@@ -0,0 +1,30 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class TrainerMapper {
```
Since this is a utility class, consider making sure that no other classes can extend or instantiate this TrainerMapper class.

MadinaBaizhanova commented on src/main/java/com/gymcrm/mapper/TrainingMapper.java:
```diff
@@ -0,0 +1,38 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class TrainingMapper {
```
Since this is a utility class, consider making sure that no other classes can extend or instantiate this TrainingMapper class.

MadinaBaizhanova commented on src/main/java/com/gymcrm/mapper/UserMapper.java:
```diff
@@ -0,0 +1,29 @@
+import com.gymcrm.dto.TrainerRequest;
+import com.gymcrm.model.User;
+
+public class UserMapper {
```
Since this is a utility class, consider making sure that no other classes can extend or instantiate this UserMapper class.

MadinaBaizhanova commented on src/main/java/com/gymcrm/model/Trainee.java:
```diff
@@ -0,0 +1,16 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class Trainee {
```
Consider replacing Integer by BigInteger in the future for the id.

MadinaBaizhanova commented on src/main/java/com/gymcrm/service/TraineeService.java:
```diff
@@ -0,0 +1,18 @@
+import java.util.List;
+import java.util.Optional;
+
+public interface TraineeService {
```
There are several methods that get repeated in Trainee, Trainer, Training, and User services. Consider grouping these repetitive methods in some base interface/abstract class.

MadinaBaizhanova commented on src/main/java/com/gymcrm/GymCLI.java:
Consider replacing System.out statements by a logger. I have once been told that having Sout real projects is a bad practice.

MadinaBaizhanova reviewed (changes requested):
Please review the comments that I left. 

Some comments actually relate to several cases, I just did not duplicate them.

Atabek04 commented on src/main/java/com/gymcrm/dao/AbstractDAO.java:
```diff
@@ -0,0 +1,54 @@
+    protected final Logger logger;
+    private final String entityName;
+
+    public AbstractDAO(Map<Integer, T> storage, Logger logger, String entityName) {
```
> Consider changing the access modifier of the AbstractDAO class constructor to 'protected' since abstract classes are only meant to be extended by their subclasses and not instantiated directly.

Yes, making AbstractDAO<T> class protected is a good idea if it is only meant to be extended by subclasses within the same package. 
However, in my case subclasses of AbstractDAO class in different packages (impl) , that's why I think it should remain public.

Atabek04 commented:
Applied review changes
- Added abstract class to Service
- Improved data validation
- Removed magic number and etc.

Please re-review the new changes and give approval for merging commits to main

MadinaBaizhanova reviewed (commented):
Hi Otabek,

Please review my comments! Should you have any questions, please let me know!

MadinaBaizhanova commented on src/main/java/com/gymcrm/util/UserUtils.java:
```diff
@@ -0,0 +1,27 @@
+        return username;
+    }
+
+    public static String generateRandomPassword() {
```
Okay, no problem for me!

Atabek04 commented on src/main/java/com/gymcrm/GymCLI.java:
Yeah, you're right about souts. 
However If I'll use logger for prompting (like: "enter your name: "), logger will show dateTime and logger level which is a little bit weird and not clean. 
Only because of that and because I'm building CLI app I'm using sout. But for sure I got you, and in the next web apps I will make sure to not use them))

maximelyashevich reviewed (approved):
Good job! Please review my comment to further improve your code quality. Approved from my side.

Regards,
Maks

Atabek04 commented on src/main/java/com/gymcrm/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,16 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GlobalExceptionHandler {
```
I saw some practices like that in Spring boot courses, where exceptions are handler automatically. But in my case it's manually, so maybe I can remove it and handle it directly in try-catch:

```java
try {
    // Some code that might throw exceptions
} catch (ResourceNotFoundException ex) {
    // GlobalExceptionHandler.handleResourceNotFoundException(ex);
    log.error("Resource not found: {}", ex.getMessage());
} catch (Exception ex) {
    // GlobalExceptionHandler.handleGeneralException(ex);
    log.error("An unexpected error occurred: {}", ex.getMessage());
}
```
